File: android/src/main/java/com/opacitylabs/capacitor/OpacityPlugin.java
package com.opacitylabs.capacitor;

import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;
import com.opacitylabs.opacitycore.OpacityCore;

@CapacitorPlugin(name = "Opacity")
public class OpacityPlugin extends Plugin {
    @PluginMethod
    public void initialize(PluginCall call) {
        String apiKey = call.getString("apiKey");
        boolean dryRun = call.getBoolean("dryRun");
        int environmentRaw = call.getInt("environment");

        OpacityCore.Environment environment = switch (environmentRaw) {
            case 0 -> OpacityCore.Environment.TEST;
            case 1 -> OpacityCore.Environment.LOCAL;
            case 2 -> OpacityCore.Environment.STAGING;
            default -> OpacityCore.Environment.PRODUCTION;
        };

        OpacityCore.initialize(apiKey, dryRun, environment);
        call.resolve(null);
    }

    @PluginMethod
    public void get(PluginCall call) {
        String name = call.getString("name");
        JSObject params = call.getObject("params");
    }
}

File: android/src/main/java/com/opacitylabs/capacitor/OpacityPlugin.kt
package com.opacitylabs.capacitor

import com.getcapacitor.Plugin
import com.getcapacitor.PluginCall
import com.getcapacitor.PluginMethod
import com.getcapacitor.annotation.CapacitorPlugin
import com.opacitylabs.opacitycore.OpacityCore
import com.opacitylabs.opacitycore.OpacityCore.initialize
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch

@CapacitorPlugin(name = "Opacity")
class OpacityPlugin : Plugin() {

    override fun load() {
        super.load()
        OpacityCore.setContext(activity)
    }

    @PluginMethod
    fun initialize(call: PluginCall) {
        val apiKey = call.getString("apiKey")
        val dryRun = call.getBoolean("dryRun")
        val environmentRaw = call.getInt("environment")
        val shouldShowErrorsInWebView = call.getBoolean("shouldShowErrorsInWebView")

        if (apiKey == null || environmentRaw == null) {
            call.reject("Invalid Arguments")
            return
        }

        val environment: OpacityCore.Environment = when (environmentRaw) {
            0 -> OpacityCore.Environment.TEST
            1 -> OpacityCore.Environment.LOCAL
            2 -> OpacityCore.Environment.STAGING
            else -> OpacityCore.Environment.PRODUCTION
        }

        initialize(apiKey, dryRun ?: false, environment, shouldShowErrorsInWebView ?: true)
        call.resolve(null)
    }

    @PluginMethod
    fun get(call: PluginCall) {
        val name = call.getString("name")
        val params = call.getObject("params", null)
        if (name == null) {
            call.reject("Name is required")
            return
        }

        CoroutineScope(Dispatchers.Main).launch {
            try {
                if (params != null) {
                    val res = OpacityCore.get(name, Converter.jsObjectToMap(params))
                    call.resolve(Converter.mapToJsObject(res))
                } else {
                    val res = OpacityCore.get(name, null)
                    call.resolve(Converter.mapToJsObject(res))
                }
            } catch (e: Exception) {
                call.reject(e.toString())
            }
        }

    }
}
